server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -955,13 +1090,26 @@ public Object[] addGeneratedValues(IndexItem item) {
                     child.name(),
                     child.path(),
                     value,
-                    Map::putIfAbsent
+                    Map::putIfAbsent,
+                    false // i.e., a user may want to set null to an object column with non-deterministic sub-cols
                 );
             }
         }
         return extendedValues.toArray();
     }
 
+    public Indexer onConflictIndexer() {
+        return this.onConflictIndexer;
+    }
+
+    public List<Reference> indexOrder() {
+        return this.indexingOrder.get();
+    }
+
+    public Object synthetic(Reference r) {
+        return synthetics.get(r.column()).computedValue;
+    }
```

Comment:
That we need to expose these and have the `TransportShardUpsertAction` re-order the result seems odd to me and is probably also expensive. Can't we create the result in the right order in the first place?